docs/src/main/paradox/user/lang/dart/eventsourced.md

```diff
@@ -0,0 +1,100 @@
+
+@@snip [eventsourced_entity.dart](/docs/src/test/dart/docs/user/eventsourced/eventsourced_entity.dart) { #add-item }
+
+This command handler also validates the command, ensuring the quantity items added is greater than zero. Invoking `ctx.fail` fails the command - this method throws so there's no need to explicitly throw an exception.
+
+## Handling events
+
+Event handlers are invoked at two points, when restoring entities from the journal, before any commands are handled, and each time a new event is emitted. An event handlers responsibility is to update the state of the entity according to the event. Event handlers are the only place where its safe to mutate the state of the entity at all.
+
+Event handlers are declared by annotating a method with `@EventHandler()`. They take a context class of type `EventContext`.
+
+Event handlers are differentiated by the type of event they handle. By default, the type of event an event handler handles will be determined by looking for a single non context parameter that the event handler takes. If for any reason this needs to be overridden, or if the event handler method doesn't take any non context parameter (because the event type may be all that needs to be known to handle the event), the type of event the handler handles can be specified using the `eventClass` parameter on the `@EventHandler()` annotation.
+
+Here's an example event handler for the `ItemAdded` event. A utility method, `convert` is also defined to assist it.
+
+@@snip [eventsourced_entity.dart](/docs/src/test/dart/docs/user/eventsourced/eventsourced_entity.dart) { #item-added }
+
+## Producing and handling snapshots
+
+Snapshots are an important optimisation for event sourced entities that may contain many events, to ensure that they can be loaded quickly even when they have very long journals. To produce a snapshot, a method annotated with `@Snapshot()` must be declared. It takes a context class of type `SnapshotContext`, and must return a snapshot of the current state in serializable form. 
+
+@@snip [eventsourced_entity.dart](/docs/src/test/dart/docs/user/eventsourced/eventsourced_entity.dart) { #snapshot }
+
+When the entity is loaded again, the snapshot will first be loaded before any other events are received, and passed to a snapshot handler. Snapshot handlers are declare by annotating a method with `@SnapshotHandler()`, and it can take a context class of type `SnapshotContext`.
+
+Multiple snapshot handlers may be defined to handle multiple different types of snapshots, the type matching is done in the same way as for events.
+
+@@snip [eventsourced_entity.dart](/docs/src/test/dart/docs/user/eventsourced/eventsourced_entity.dart) { #handle-snapshot }
+
+## Registering the entity
+
+Once you've created your entity, you can register it with the `cloudState` server, by invoking the `registerEventSourcedEntity` function. In addition to passing the protobuf qualified package name, and your entity class.
```

Comment:
`registerEventSourcedEntity` is mentioned here but not be seen in the code following this paragraph. Its mentioned in the page before but kind of lacks presence here.